parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java

```diff
@@ -0,0 +1,621 @@
+    // Shift the just-written field data to make room for the header section.
+    System.arraycopy(writeBuffer, start, writeBuffer, start + headerSize, dataSize);
+    writePos += headerSize;
+    writeBuffer[start] = VariantUtil.arrayHeader(largeSize, offsetSize);
+    VariantUtil.writeLong(writeBuffer, start + 1, size, sizeBytes);
+    int offsetStart = start + 1 + sizeBytes;
+    for (int i = 0; i < size; ++i) {
+      VariantUtil.writeLong(
+          writeBuffer,
+          offsetStart + i * offsetSize,
+          arrayBuilder.offsets().get(i),
+          offsetSize);
+    }
+    VariantUtil.writeLong(writeBuffer, offsetStart + size * offsetSize, dataSize, offsetSize);
+    return this;
+  }
+
+  /**
+   * Adds a key to the Variant dictionary. If the key already exists, the dictionary is unmodified.
+   * @param key the key to add
+   * @return the id of the key
+   */
+  int addKey(String key) {
+    return dictionary.computeIfAbsent(key, newKey -> {
+      int id = dictionaryKeys.size();
+      dictionaryKeys.add(newKey.getBytes(StandardCharsets.UTF_8));
+      return id;
+    });
+  }
+
+  /**
+   * @return the current write position of the variant builder
+   */
+  int writePos() {
+    return writePos;
+  }
+
+  /**
+   * Class to store the information of a Variant object field. We need to collect all fields of
+   * an object, sort them by their keys, and build the Variant object in sorted order.
+   */
+  static final class FieldEntry implements Comparable<FieldEntry> {
+    final String key;
+    final int id;
+    final int offset;
+
+    FieldEntry(String key, int id, int offset) {
+      this.key = key;
+      this.id = id;
+      this.offset = offset;
+    }
+
+    FieldEntry withNewOffset(int newOffset) {
+      return new FieldEntry(key, id, newOffset);
+    }
+
+    @Override
+    public int compareTo(FieldEntry other) {
+      return key.compareTo(other.key);
+    }
+  }
+
+  private void checkCapacity(int additionalBytes) {
+    int requiredBytes = writePos + additionalBytes;
+    if (requiredBytes > writeBuffer.length) {
+      // Allocate a new buffer with a capacity of the next power of 2 of `requiredBytes`.
+      int newCapacity = Integer.highestOneBit(requiredBytes);
+      newCapacity = newCapacity < requiredBytes ? newCapacity * 2 : newCapacity;
+      byte[] newValue = new byte[newCapacity];
+      System.arraycopy(writeBuffer, 0, newValue, 0, writePos);
+      writeBuffer = newValue;
+    }
+  }
+
+  private int getMinIntegerSize(int value) {
+    assert value >= 0;
+    if (value <= VariantUtil.U8_MAX) {
+      return VariantUtil.U8_SIZE;
+    }
+    if (value <= VariantUtil.U16_MAX) {
+      return VariantUtil.U16_SIZE;
+    }
+    if (value <= VariantUtil.U24_MAX) {
+      return VariantUtil.U24_SIZE;
+    }
+    return VariantUtil.U32_SIZE;
+  }
+
+  /** The buffer for building the Variant value. The first `writePos` bytes have been written. */
+  private byte[] writeBuffer = new byte[128];
```

Comment:
If this is trying to avoid allocations, then I think this should start at a larger size. It would make sense to me to start a 1k or 2k, considering this is probably reused and we don't want to reallocate very often.